.github/workflows/ci.yml

```diff
@@ -11,29 +11,29 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout repository
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
 
       - name: Set up Python
-        uses: actions/setup-python@v4
+        uses: actions/setup-python@v5
 
       - name: Run pre-commit
-        run: pipx run -- hatch run lint:run
+        run: python3 -m pip install hatch && hatch run lint:run
 
   test:
     strategy:
       matrix:
         os: [ubuntu-latest, macos-latest]
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.9", "3.10", "3.11", "3.12", "3.13", "pypy-3.10"]
 
     runs-on: ${{ matrix.os }}
     steps:
       - name: Checkout repository
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
 
       - name: Set up Python
-        uses: actions/setup-python@v4
+        uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.python-version }}
 
       - name: Run tests
-        run: pipx run -- hatch run test:run
+        run: python3 -m pip install hatch && hatch run test:run
```

Comment:
**n.b.** Due to `pipx run --` usage, the cpython version being used was always `python 3.10` despite what was set by `actions/setup-python@v5`.